🚀 1단계 - 구간 추가 요구사항 반영

hhhhhwi opened the pull request:
안녕하세요 용주님 🙇 
늦게나마 인사드리네요..! 새해 복 많이 받으세요 🐉  
이제서야 1단계 PR을 올립니다 ㅠ

이번 미션에서는 테스트 시나리오 작성과 해당 시나리오대로 작성한 인수테스트대로 구현하는 방향에 집중해서 미션에 임했습니다..!

요구사항을 추가하다보니.. line과 section의 책임이 혼동되는 느낌이 있어 좀 고민이 되었습니다 ㅠ

지난 미션에서 주셨던 피드백도 함께 적용하려고 했는데요..!
혹시나 제가 빼먹은 부분이 있으면 이번 피드백 때 다시 한 번 적용해보도록 하겠습니다!

hhhhhwi committed:
refactor: 무분별한 객체 생성을 방지하기위하 기본 생성자의 접근 범위를 protected로 수정

hhhhhwi committed:
docs: readme 작성

hhhhhwi committed:
refactor: 불필요한 주석 수정

hhhhhwi committed:
test: 구간 추가 요구사항 반영 인수 테스트 작성

hhhhhwi committed:
feat: 구간 추가 요구사항 반영

hhhhhwi committed:
refactor: if - else 메소드로 분리

hhhhhwi committed:
refactor: 사용하지 않는 import 구문 삭제

hhhhhwi committed:
docs: readme 수정

testrace commented on README.md:
```diff
@@ -1,2 +1,57 @@
+- [x] api 구현
+- [ ] 단위테스트
+  - [ ] 객체 단위 테스트
+  - [ ] 서비스 단위 테스트
```
요구사항을 readme에 잘 정리해 주셨네요 👍 
단위 테스트도 추가해 보시면 좋을 것 같아요!

testrace commented on src/main/java/nextstep/subway/entity/Line.java:
```diff
@@ -22,7 +21,7 @@ public class Line {
 	private Sections sections;
 
-	public Line() {
+	protected Line() {
```
protected 👍

testrace commented on src/main/java/nextstep/subway/entity/Line.java:
```diff
@@ -70,18 +69,29 @@ public void setUpdateInfo(String name, String color) {
+			return;
+		}
+
+		if(isEndStation(section.getUpStationId())) {
```
```suggestion
		if (isEndStation(section.getUpStationId())) {
```

testrace commented on src/main/java/nextstep/subway/entity/Line.java:
```diff
@@ -70,18 +69,29 @@ public void setUpdateInfo(String name, String color) {
 	}
 
 	public void addSection(Section section) {
```
>요구사항을 추가하다보니.. line과 section의 책임이 혼동되는 느낌이 있어 좀 고민이 되었습니다 ㅠ

노선 없는 구간은 존재할 수 없으니 구간 관리라는 기능은 line의 책임으로 봐도 이상하지 않다고 생각합니다 😃 
다만 모든 코드가 line 내에 있으면 클래스가 커져 유지보수성이 떨어지게 됩니다
line의 일부 책임을 sections에게 위임하면 클래스를 작게 유지할 수 있어 유지보수성을 높일 수 있습니다 
sections 또한 너무 많은 책임을 가지고 있다면 클래스를 분리해서 위임할 수도 있겠네요
sections 클래스가 도출되어 있으니 '구간 추가' 책임은 sections에게 있다고 보시면 좋을 것 같아요.

testrace commented on src/main/java/nextstep/subway/entity/Line.java:
```diff
@@ -112,4 +122,19 @@ public boolean isStartStation(Long stationId) {
+		this.sections.addSection(section);
+	}
+
+	private void addMidSection(Section section) {
+		Section upSection = sections.getSectionByUpStationId(section.getUpStationId());
+		if(upSection.getDistance() <= section.getDistance()) {
+			throw new IllegalArgumentException("등록 구간의 길이는 기존 구간의 길이보다 크거나 같을 수 없습니다.");
+		}
+		sections.addSection(section);
+		sections.addSection(new Section(this, section.getDownStationId(), upSection.getDownStationId(), upSection.getDistance() - section.getDistance()));
+		sections.removeSection(upSection);
+	}
```
이 메서드를 포함해 구간과 관련된 메서드들은 모두 sections로 이동해 보셔도 좋을 것 같아요

hhhhhwi replied:
Line 객체에 distance와 startStationId, EndStationId가 있다보니 완전한 분리가 쉽지는 않더라구요..! 
위 변수들이 line의 책임이 맞을까? 라는 고민이 들기도 했습니다😢
일단 sections가 변경되는 내용만 Sections 클래스에 옮기는 방향으로 수정해보았습니다!

testrace commented on src/main/java/nextstep/subway/entity/Sections.java:
```diff
@@ -48,6 +48,13 @@ public Section getSectionByDownStationId(Long downStationId) {
 				.orElseThrow(EntityNotFoundException::new);
 	}
 
+	public Section getSectionByUpStationId(Long upStationId) {
+		return sections.stream()
+				.filter(x-> upStationId.equals(x.getUpStationId()))
+				.findAny()
+				.orElseThrow(EntityNotFoundException::new);
+	}
```
단위 테스트가 추가되면 좋을 것 같아요 😃

testrace commented on src/test/java/nextstep/subway/acceptance/subway/SectionAcceptanceTest.java:
```diff
@@ -12,42 +13,103 @@
+		// then
+		assertThat(response.statusCode()).isEqualTo(HttpStatus.CREATED.value());
+
+		List<SectionResponse> sections = 구간_조회_요청(노선).jsonPath().getList("", SectionResponse.class);
+		assertTrue(sections.stream()
+				.anyMatch(section -> 종로3가역.equals(section.getUpStationId()) && 종각역.equals(section.getDownStationId())));
+		assertTrue(sections.stream()
+				.anyMatch(section -> 종각역.equals(section.getUpStationId()) && 시청역.equals(section.getDownStationId()) && section.getDistance() == 2));
```
assert 내부에서 stream 등의 코드가 있으면 코드를 이해하는데 한 번 더 생각을 해야해서 직관적이지 못한 것 같아요
빠르게 테스트의 의도를 파악하기 위해 노선 관리 인수 테스트처럼 직관적으로 작성하면 어떨까요?
예시는 참고만 해주세요
```java
assertThat(노선의_역_목록).containsExactly(종로3가역, 종각역, 시청역);

```

hhhhhwi replied:
아예 검증 로직 자체를 메소드로 빼는 식으로 수정해봤습니다..!

pr올리고 다시 보니제안해주신 것 처럼 검증 조건문만 따로 빼는 게 더 깔끔해보이네요..🤔

testrace commented on src/test/java/nextstep/subway/acceptance/subway/SectionAcceptanceTest.java:
```diff
@@ -59,37 +121,53 @@ void createSectionTest() {
+		실패시_코드값_메시지_검증(response, HttpStatus.BAD_REQUEST.value(),"등록 구간의 역들이 모두 노선에 존재하지 않습니다.");
+	}
+
+	/**
+	 * Given 지하철 노선을 생성한다.
+	 * Given 해당 지하철 노선은
+	 * 		상행역 : 종로3가역 / 하행역 : 시청역 / 길이 : 6 인 구간이 존재한다.
+	 * When 생성한 지하철 노선에
+	 * 		상행역 : 종로3가역 / 하행역 : 종각역 / 길이가 6 이상인 구간을 등록하면
+	 * Then 등록되지 않고 코드값 400 (Bad Request) 을 반환한다.
+	 */
```
모든 인수 조건이 인상적이에요 인수 조건을 상당히 잘 도출하시는 것 같아요 👍 
인수 조건은 개발자만의 관점이 아닌 사용자 관점이니 400 http statu code를 명시하지 않아도 괜찮을 것 같은데 어떻게 생각하시는지 궁금합니다 🤔

hhhhhwi replied:
코드 값도 응답 값이라고 생각해서 해당 시나리오를 작성했는데 너무 제 입장만 생각했던 것 같습니다..! 😂 리턴 메시지 값 검증으로 시나리오 변경 해봤습니다!

testrace reviewed (changes requested):
안녕하세요 휘연님 😃 
1단계 구현 잘해주셨네요 👍 
질문에 대한 답변과 코멘트 확인해 주시고 다시 리뷰 요청 부탁드려요

hhhhhwi committed:
refactor: Java Convention에 맞게 수정

hhhhhwi committed:
refactor: 검증 내용을 메소드로 분리, 실패 인수 조건 검증 시나리오 변경
- 빠른 테스트 의도 파악을 위해 직관적인 메소드 명을 사용하여 분리
- 사용자 관점에서의 인수 테스트를 위해 http status 코드 값이 아닌 리턴 메시지 값을 통해 검증하도록 시나리오 수정

hhhhhwi committed:
refactor: 구간 추가 및 제거에 대한 책임을 Sections로 이동

hhhhhwi committed:
test: Sections 단위 테스트 구현

testrace commented on src/main/java/nextstep/subway/entity/Line.java:
```diff
@@ -119,22 +114,17 @@ public boolean isStartStation(Long stationId) {
-	private void addStartOrEndSection(Section section) {
-		this.distance += section.getDistance();
-		this.sections.addSection(section);
+	private void updateStartStation(Long stationId, int distance) {
+		this.startStationId = stationId;
+		this.distance += distance;
```
https://github.com/next-step/atdd-subway-path/pull/736#discussion_r1485736038
>Line 객체에 distance와 startStationId, EndStationId가 있다보니 완전한 분리가 쉽지는 않더라구요..!
위 변수들이 line의 책임이 맞을까? 라는 고민이 들기도 했습니다😢

Line의 distance, startStationId, endStationId 모두 구간과 관련된 필드이니
Sections를 통해 구하는 방법도 고려해 보시면 좋을 것 같아요 😃

testrace commented on src/main/java/nextstep/subway/entity/Sections.java:
```diff
@@ -23,17 +23,32 @@ public void addSection(Section section) {
 	}
 
-	public void removeSection(Section section) {
+	public void addMidSection(Line line, Section section) {
```
Sections 의 책임이 명확해진 것 같아요 👍

testrace commented on src/test/java/nextstep/subway/unit/SectionsTest.java:
```diff
@@ -0,0 +1,60 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
+
+class SectionsTest {
```
피드백 반영 👍 
다음 단계에서는 tdd 사이클을 지켜서 진행해 보시면 좋을 것 같아요

testrace commented on src/test/java/nextstep/subway/unit/SectionsTest.java:
```diff
@@ -0,0 +1,60 @@
+    private static final Long 종로3가역 = 1L;
+    private static final Long 시청역 = 2L;
+    private static final Long 서울역 = 3L;
+    private static final Long 종각역 = 4L;
+    private Line 노선;
```
```suggestion
    private static final Long 종각역 = 4L;
    
    private Line 노선;
```

testrace commented on src/test/java/nextstep/subway/acceptance/subway/SectionAcceptanceTest.java:
지난 피드백에서 코멘트 드리는걸 놓친 부분이 있는데요
전체 테스트를 수행할 때 성공하지 못하는 케이스가 존재해요 😢 
<img width="386" alt="image" src="https://github.com/next-step/atdd-subway-path/assets/8064743/67f1bf8f-27d8-4133-b3a5-d6947650a008">

testrace reviewed (approved):
피드백 반영 잘해주셨네요 👍 
몇 가지 코멘트 남겨두었는데 확인해 주시고 다음 단계 진행해 주세요!
